File: java/core/src/core/crypt/PBE.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */

package core.crypt;

import org.bc.crypto.PBEParametersGenerator;
import org.bc.crypto.digests.SHA256Digest;
import org.bc.crypto.generators.PKCS5S2ParametersGenerator;
import org.bc.crypto.params.KeyParameter;

import core.callback.Callback;
import core.callback.CallbackDefault;
import core.crypt.Cryptor;
import core.crypt.CryptorAES;
import core.exceptions.CryptoException;
import core.util.Arrays;
import core.util.Base64;
import core.util.LogNull;
import core.util.LogOut;

/**
 * http://stackoverflow.com/questions/992019/java-256-bit-aes-password-based-encryption
 */
public class PBE extends Cryptor
{
	LogNull log = new LogNull(PBE.class);
	
	public byte[] key = null;
	Cryptor cryptorAES;
	
    public static final int DEFAULT_ITERATIONS = 131072;
    public static final int SHORT_ITERATIONS = 32768;
    public static final int DEFAULT_KEYLENGTH = 256;
    
    /**
     * bytes used to salt the key (set before making an instance)
     */
    public static final byte[] DEFAULT_SALT_0 = {
        (byte)0xc8, (byte)0x73, (byte)0x41, (byte)0x8c,
        (byte)0x7e, (byte)0xd8, (byte)0xee, (byte)0x89
    };
    
    /**
     * bytes used to salt the key (set before making an instance)
     */
    public static final byte[] DEFAULT_SALT_1 = {
        (byte)0x12, (byte)0x53, (byte)0x14, (byte)0xbb,
        (byte)0x7e, (byte)0x97, (byte)0xce, (byte)0x55
    };
    
    /**
     * bytes used to salt the key (set before making an instance)
     */
    public static final byte[] DEFAULT_SALT_2 = {
        (byte)0x0a, (byte)0x48, (byte)0x33, (byte)0xfe,
        (byte)0xa7, (byte)0xc2, (byte)0x2c, (byte)0x71
    };

    public static final byte[][] DEFAULT_SALT = { DEFAULT_SALT_0, DEFAULT_SALT_1, DEFAULT_SALT_2 };

    public PBE ()
    {
    	
    }
    public PBE (String password, byte[] salt, int iterationCount, int keyLength) throws CryptoException
    {
    	generate(password,salt, iterationCount, keyLength);
    }
    
    public void generate (String password, byte[] salt, int iterationCount, int keyLength)
    	throws CryptoException 
	{
    	key = PBEPlatform.generate(password,  salt, iterationCount, keyLength);
		cryptorAES = new CryptorAES (key);
		
		log.debug("computed", new String(Base64.encode(key)), " using ", new String(password),"salt",new String(Base64.encode(salt)), iterationCount, keyLength);
	}
	
    public Callback generate_ (String password, byte[] salt, int iterationCount, int keyLength)
    {
    	return 
	    	new CallbackDefault(password, salt, iterationCount, keyLength) {
				public void onSuccess(Object... arguments) throws Exception {
					String password = V(0);
					byte[] salt = V(1);
					int iterationCount = (Integer)V(2);
					int keyLength = (Integer)V(3);
					
					generate(password, salt, iterationCount, keyLength);
			}
		};
	}
    
    public byte[] encrypt(byte[] clearText) throws CryptoException 
    {
    	return cryptorAES.encrypt(clearText);
    }
    	
    public byte[] decrypt(byte[] encrypted) throws CryptoException
    {
    	return cryptorAES.decrypt(encrypted);
    }
}

File: cpp/core/mailiverse/core/crypt/PBE.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "PBE.h"
#include "CryptorAES.h"
#include "mailiverse/utilities/Log.h"

#include <botan/botan.h>
#include <botan/pbkdf2.h>
#include <botan/sha2_32.h>

using namespace mailiverse::core::crypt;
using namespace mailiverse::core;
using namespace mailiverse::utilities;
using namespace mailiverse;

const Byte DEFAULT_SALT_CRYPTOR_INITIALIZER[] = { 
	0xc8, 0x73, 0x41, 0x8c,
	0x7e, 0xd8, 0xee, 0x89
} ;

const Salt PBE::DEFAULT_SALT_CRYPTOR = toBlockFromInitializer (DEFAULT_SALT_CRYPTOR_INITIALIZER);

const Byte DEFAULT_SALT_VERIFIER_INITIALIZER[] = {
	0x12, 0x53, 0x14, 0xbb,
	0x7e, 0x97, 0xce, 0x55
} ;

const Salt PBE::DEFAULT_SALT_VERIFIER = toBlockFromInitializer (DEFAULT_SALT_VERIFIER_INITIALIZER);


PBE::PBE (const std::string &password, const Salt &salt, int iterationCount, int keyLength)
{
	Botan::Algorithm_Factory& af = Botan::global_state().algorithm_factory();
	Botan::MessageAuthenticationCode *prf = af.make_mac("HMAC(SHA-256)");
	Botan::PKCS5_PBKDF2 kdf(prf);
	
	Botan::OctetString key = 
		kdf.derive_key(keyLength/8, password, salt.data(), salt.size(), iterationCount);
	
	secretKey.assign(key.begin(), key.end());
	cryptor = new CryptorAES(secretKey);
	
	LogDebug(mailiverse::core::crypt::PBE, "derived: " << toString(toBlockBase64(secretKey)) << " using " << toString(toBlockBase64(salt)) << " password " << password);
}

PBE::~PBE ()
{
}

Block PBE::encrypt(const Block::Atom *begin, const Block::Atom *end)
{
	return cryptor->encrypt(begin, end);
}

Block PBE::decrypt(const Block::Atom *begin, const Block::Atom *end)
{
	return cryptor->decrypt(begin, end);
}
